### PR TITLE
Implement node realization

### PR DESCRIPTION
**Pull request by FranckLecuyer**

### CODE REVIEW

File: src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java [hunk 1]
```diff
@@ -745,4 +946,201 @@ private void createTwoWindingsTransformer(Network network, VoltageLevel voltageL
...
+            null,
+            modificationRepository,
+            equipmentInfosService,
+            true,
+            true);
+        ReporterModel reporter = new ReporterModel("Realizaton", "Realization");
+
+        modificationRepository.getModificationsEntities(realizationInfos.getModifications()).forEach(modificationEntity -> {
+            ModificationType type = ModificationType.valueOf(modificationEntity.getType());
+            switch (type) {
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java:1062]: Need to have an object oriented approach ?
Review the design with a `Modification` interface and a class for each type which implements the interface
and a factory to create these classes
```
interface Modification {
    List<EquipmenModificationInfos> apply();
    
    void store();
}
```
1. For a modification from the front we create the modification then apply it and store it
2. For a realization get the modification from the database then  apply it only

The listener must be used only for the application.

**Comment by SlimaneAmar** [comment]
See comments for the design
I think also all the `transactional` annotations are not necessary

**Review by SlimaneAmar — CHANGES_REQUESTED** [review]

File: src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java [hunk 2]
```diff
@@ -745,4 +946,201 @@ private void createTwoWindingsTransformer(Network network, VoltageLevel voltageL
...
+        List<EquipmenModificationInfos> allModificationsInfos = new ArrayList<>();
+        NetworkStoreListener listener = NetworkStoreListener.create(network,
+            networkUuid,
+            realizationInfos.getDestinationVariantId(),
+            null,
+            modificationRepository,
+            equipmentInfosService,
+            true,
+            true);
+        ReporterModel reporter = new ReporterModel("Realizaton", "Realization");
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java:1058]: Typo

Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java:1058]: Corrected

File: src/main/java/org/gridsuite/modification/server/service/RealizationWorkerService.java [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+    private Set<String> realizationRequests = Sets.newConcurrentHashSet();
+
+    @Autowired
+    private StreamBridge resultRealizationMessagePublisher;
+
+    public RealizationWorkerService(NetworkModificationService networkModificationService,
+                                    ObjectMapper objectMapper,
+                                    RealizationStoppedPublisherService stoppedPublisherService) {
+        this.networkModificationService = Objects.requireNonNull(networkModificationService);
```
[the remark this answers is not in the document]
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/modification/server/service/RealizationWorkerService.java:72]: Done

File: src/main/java/org/gridsuite/modification/server/service/RealizationExecContext.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    private static String getNonNullHeader(MessageHeaders headers, String name) {
+        String header = (String) headers.get(name);
+        if (header == null) {
+            throw new PowsyblException("Header '" + name + "' not found");
+        }
+        return header;
+    }
+
+    public static RealizationExecContext fromMessage(Message<String> message, ObjectMapper objectMapper) {
+        Objects.requireNonNull(message);
```
[the remark this answers is not in the document]
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/modification/server/service/RealizationExecContext.java:59]: Done

File: src/main/java/org/gridsuite/modification/server/service/RealizationExecContext.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+public class RealizationExecContext {
+
+    private UUID networkUuid;
+
+    private RealizationInfos realizationInfos;
+
+    private String receiver;
+
+    public RealizationExecContext(UUID networkUuid, RealizationInfos realizationInfos, String receiver) {
+        this.networkUuid = Objects.requireNonNull(networkUuid);
```
[the remark this answers is not in the document]
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/modification/server/service/RealizationExecContext.java:33]: Done

File: src/main/java/org/gridsuite/modification/server/service/RealizationCancelContext.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    public RealizationCancelContext(String receiver) {
+        this.receiver = Objects.requireNonNull(receiver);
+    }
+
+    public String getReceiver() {
+        return receiver;
+    }
+
+    public static RealizationCancelContext fromMessage(Message<String> message) {
+        Objects.requireNonNull(message);
```
[the remark this answers is not in the document]
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/modification/server/service/RealizationCancelContext.java:31]: Done

File: src/main/java/org/gridsuite/modification/server/service/RealizationCancelContext.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+/**
+ * @author Franck Lecuyer <franck.lecuyer at rte-france.com>
+ */
+public class RealizationCancelContext {
+
+    private final String receiver;
+
+    public RealizationCancelContext(String receiver) {
+        this.receiver = Objects.requireNonNull(receiver);
```
[the remark this answers is not in the document]
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/modification/server/service/RealizationCancelContext.java:23]: Done

File: src/main/java/org/gridsuite/modification/server/service/RealizationStoppedPublisherService.java [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+    public void publishCancel(String receiver) {
+        publish(receiver, CANCEL_MESSAGE);
+    }
+
+    public void publishFail(String receiver, String causeMessage) {
+        publish(receiver, FAIL_MESSAGE + " : " + causeMessage);
+    }
+
+    public void publish(String receiver, String stopMessage) {
```
[the remark this answers is not in the document]
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/modification/server/service/RealizationStoppedPublisherService.java:41]: Done

File: src/main/java/org/gridsuite/modification/server/NetworkModificationController.java [hunk 9]
```diff
@@ -140,4 +142,21 @@ public NetworkModificationController(NetworkModificationService networkModificat
...
+    @PostMapping(value = "/networks/{networkUuid}/realization", produces = MediaType.APPLICATION_JSON_VALUE)
+    @Operation(summary = "Realize a network variant")
+    @ApiResponses(value = {@ApiResponse(responseCode = "200", description = "The realization has been done")})
+    public ResponseEntity<Mono<Void>> realizeVariant(@Parameter(description = "Network UUID") @PathVariable("networkUuid") UUID networkUuid,
+                                                     @Parameter(description = "Receiver") @RequestParam(name = "receiver", required = false) String receiver,
+                                                     @RequestBody RealizationInfos realizationInfos) {
+        return ResponseEntity.ok().body(networkModificationService.realizeVariant(networkUuid, realizationInfos, receiver));
+    }
+
+    @PutMapping(value = "/realization/stop", produces = APPLICATION_JSON_VALUE)
```
[the remark this answers is not in the document]
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/modification/server/NetworkModificationController.java:155]: Done

File: src/main/java/org/gridsuite/modification/server/NetworkModificationController.java [hunk 10]
```diff
@@ -140,4 +142,21 @@ public NetworkModificationController(NetworkModificationService networkModificat
                                                                         @RequestParam(value = "group", required = false) UUID groupUuid) {
         return ResponseEntity.ok().body(networkModificationService.deleteEquipment(networkUuid, variantId, groupUuid, equipmentType, equipmentId));
     }
+
+    @PostMapping(value = "/networks/{networkUuid}/realization", produces = MediaType.APPLICATION_JSON_VALUE)
```
[the remark this answers is not in the document]
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/modification/server/NetworkModificationController.java:146]: Done

File: src/main/java/org/gridsuite/modification/server/service/RealizationWorkerService.java [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+        if (receiver != null) {
+            futures.put(receiver, future);
+        }
+        if (receiver != null && cancelRealizationRequests.get(receiver) != null) {
+            return Mono.empty();
+        } else {
+            LOGGER.info("Starting realization on variant : {}", realizationInfos.getDestinationVariantId());
+            try {
+                return Mono.fromCompletionStage(future);
+            } catch (Exception exc) {
```
[the remark this answers is not in the document]
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/modification/server/service/RealizationWorkerService.java:95]: Removed

File: src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java [hunk 12]
```diff
@@ -745,4 +946,201 @@ private void createTwoWindingsTransformer(Network network, VoltageLevel voltageL
...
+
+    public Mono<Network> cloneNetworkVariant(UUID networkUuid, String originVariantId, String destinationVariantId) {
+        return Mono.fromCallable(() -> {
+            Network network;
+            try {
+                network = networkStoreService.getNetwork(networkUuid);
+            } catch (PowsyblException e) {
+                throw new NetworkModificationException(NETWORK_NOT_FOUND, networkUuid.toString());
+            }
+            String startingVariant = StringUtils.isEmpty(originVariantId) ? VariantManagerConstants.INITIAL_VARIANT_ID : originVariantId;
```
[the remark this answers is not in the document]
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java:958]: Done

File: src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java [hunk 13]
```diff
@@ -212,4 +212,9 @@ public EquipmentCreationEntity createTwoWindingsTransformerEntity(String id, Str
     public EquipmentDeletionEntity createEquipmentDeletionEntity(String equipmentId, String equipmentType) {
         return new EquipmentDeletionEntity(equipmentId, equipmentType);
     }
+
+    @Transactional(readOnly = true)
```
[the remark this answers is not in the document]
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java:216]: Removed

File: src/main/java/org/gridsuite/modification/server/service/RealizationWorkerService.java [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+            .flatMap(message -> {
+                RealizationExecContext execContext = RealizationExecContext.fromMessage(message, objectMapper);
+                realizationRequests.add(execContext.getReceiver()); // receiver is the node uuid to realize
+
+                RealizationInfos realizationInfos = execContext.getRealizationInfos();
+
+                return networkModificationService.cloneNetworkVariant(execContext.getNetworkUuid(),
+                                                                      realizationInfos.getOriginVariantId(),
+                                                                      realizationInfos.getDestinationVariantId())
+                    .flatMap(network -> execRealizeVariant(network, execContext.getNetworkUuid(), execContext.getReceiver(), realizationInfos))
```
[the remark this answers is not in the document]
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/modification/server/service/RealizationWorkerService.java:113]: Done

File: src/main/java/org/gridsuite/modification/server/service/RealizationWorkerService.java [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+            }
+        }
+    }
+
+    @Bean
+    public Consumer<Flux<Message<String>>> consumeRealize() {
+        return f -> f.log(CATEGORY_BROKER_INPUT, Level.FINE)
+            .flatMap(message -> {
+                RealizationExecContext execContext = RealizationExecContext.fromMessage(message, objectMapper);
+                realizationRequests.add(execContext.getReceiver()); // receiver is the node uuid to realize
```
[the remark this answers is not in the document]
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/modification/server/service/RealizationWorkerService.java:106]: Done

File: src/main/java/org/gridsuite/modification/server/service/RealizationWorkerService.java [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+    public RealizationWorkerService(NetworkModificationService networkModificationService,
+                                    ObjectMapper objectMapper,
+                                    RealizationStoppedPublisherService stoppedPublisherService) {
+        this.networkModificationService = Objects.requireNonNull(networkModificationService);
+        this.objectMapper = Objects.requireNonNull(objectMapper);
+        this.stoppedPublisherService = Objects.requireNonNull(stoppedPublisherService);
+    }
+
+    private Mono<List<EquipmenModificationInfos>> execRealizeVariant(Network network, UUID networkUuid, String receiver, RealizationInfos realizationInfos) {
+        if (receiver != null && cancelRealizationRequests.get(receiver) != null) {
```
[the remark this answers is not in the document]
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/modification/server/service/RealizationWorkerService.java:78]: Done

File: src/main/java/org/gridsuite/modification/server/service/RealizationWorkerService.java [hunk 17]
```diff
@@ -0,0 +1,178 @@
...
+
+                            sendResultRealizationMessage(sendMessage);
+                            LOGGER.info("Realization complete on node '{}'", execContext.getReceiver());
+                        } else {  // result not available : stop realization request
+                            if (cancelRealizationRequests.get(execContext.getReceiver()) != null) {
+                                stoppedPublisherService.publishCancel(execContext.getReceiver());
+                            }
+                        }
+                    })
+                    .onErrorResume(throwable -> {
```
[the remark this answers is not in the document]
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/modification/server/service/RealizationWorkerService.java]: Done

**Review by SlimaneAmar — APPROVED** [review]

**Pull request merged**